src/Canon.hs

```diff
@@ -97,35 +102,33 @@ instance Monoid LoadReport where
       , requestBuildErrorCount = Just 0
       }
 
-runLoadTillGivenTime :: POSIXTime -> Int -> Int -> SB.SessionTemplate -> LoadReport -> IO LoadReport
-runLoadTillGivenTime intialTime numberOfParallelThreads totalTimeToRun sessionTemplate acc = do
-  currentTime <- getPOSIXTime
-  if (Utils.fromDiffTimeToSeconds $ currentTime - intialTime) < Utils.toPico totalTimeToRun
-    then do
-      res <- loadRunner numberOfParallelThreads sessionTemplate
-      runLoadTillGivenTime intialTime numberOfParallelThreads totalTimeToRun sessionTemplate (acc <> res)
-    else do
-      pure acc
-
 loadRunner :: Int -> SB.SessionTemplate -> IO LoadReport
 loadRunner sessionCount sessionTemplate = do
-  requestsForSession <- makeSessions sessionCount sessionTemplate
-  response <- withLatency $! runRequestParallely requestsForSession
+  response <- withLatency $! runRequestParallely sessionCount sessionTemplate
   return $ generateReport (fst response) (snd response)
 
-makeSessions :: Int -> SB.SessionTemplate ->  IO [SB.NormalisedSession]
-makeSessions cnt sessionTemplate = sequence $! generate <$> [1..cnt]
-  where
-    generate _ = SB.generateNewSession sessionTemplate
+runRequestParallely :: Int -> SB.SessionTemplate -> IO [[ResponseAndLatency]]
+runRequestParallely sessionCount sessionTemplate = do
+  forConcurrently [1..sessionCount] (\_ -> do 
+    responseTimeout <- Utils.responseTimeoutInSeconds <$> Ref.readIORef loadTestConfig
+    manager <- Client.newManager $ tlsManagerSettings {Client.managerResponseTimeout = Client.responseTimeoutMicro $ Utils.toMicroFromSec responseTimeout}
+    runSessionForever sessionTemplate manager [])
+
 
-runRequestParallely :: [SB.NormalisedSession] -> IO [[ResponseAndLatency]]
-runRequestParallely normalSessions = do
-  mapConcurrently runRequestSeqentially normalSessions
+runSessionForever :: SB.SessionTemplate -> Client.Manager -> [ResponseAndLatency] -> IO [ResponseAndLatency]
+runSessionForever sessionTemplate manager acc = do
+  startTime <- Ref.readIORef startTimeRef
```

Comment:
Why is this required. Can you move this outside. 

Also, can we not have this check in every run. Have it outside in a different loop with check on the time and update the Ref to True.